File: api/src/test/java/com/ford/internalprojects/peoplemover/auth/AuthServiceTest.java
/*
 * Copyright (c) 2019 Ford Motor Company
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ford.internalprojects.peoplemover.auth;

import org.junit.Before;
import org.junit.Test;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

public class AuthServiceTest {

    private RestTemplate authRestTemplate;

    private AuthService authService;

    private String fakeAccessCode = "CODE_ABC123";
    private String fakeAccessToken = "TOKEN_123";

    private String fakeClientId = "CLIENT_ID_123";
    private String fakeClientSecret = "CLIENT_SECRET_123";

    @Before
    public void setup() {

        authRestTemplate = mock(RestTemplate.class);

        AuthQuestAccessTokenRequest requestBody = AuthQuestAccessTokenRequest.builder()
                .access_code(fakeAccessCode)
                .client_id(fakeClientId)
                .client_secret(fakeClientSecret)
                .build();

        when(authRestTemplate.postForObject("/api/oauth/access_token", requestBody, AuthQuestAccessTokenResponse.class))
                .thenReturn(AuthQuestAccessTokenResponse.builder()
                        .user_id("USER_ID123")
                        .access_token(fakeAccessToken)
                        .build());

        authService = new AuthService(fakeClientId, fakeClientSecret, authRestTemplate);
    }

    @Test
    public void shouldReturnAnAuthTokenForAValidAccessCode() {

        AccessTokenResponse response = authService.getAccessToken(fakeAccessCode);

        assertThat(response.getAccessToken()).isEqualTo(fakeAccessToken);
    }

    @Test
    public void shouldReturn200IfAccessTokenIsValid() {

        var validateTokenRequest = new ValidateTokenRequest("123");

        var authQuestValidateAccessTokenRequest = AuthQuestValidateAccessTokenRequest.builder().access_token("123").build();

        when(authRestTemplate.postForEntity("/api/oauth/access_token/validate", authQuestValidateAccessTokenRequest, AuthQuestJWT.class))
                .thenReturn(ResponseEntity.ok().build());

        assertThat(authService.validateAccessToken(validateTokenRequest)).isEqualTo(ResponseEntity.ok().build());

        verify(authRestTemplate).postForEntity("/api/oauth/access_token/validate", authQuestValidateAccessTokenRequest, AuthQuestJWT.class);
    }

    @Test
    public void shouldReturnRefreshedTokenForAValidAccessToken(){
        AuthQuestRefreshTokenRequest request = AuthQuestRefreshTokenRequest.builder()
                .client_secret(fakeClientSecret)
                .client_id(fakeClientId)
                .access_token(fakeAccessToken)
                .build();

        AuthQuestAccessTokenResponse response = AuthQuestAccessTokenResponse.builder().access_token("newToken").build();
        when(authRestTemplate.postForObject("/api/oauth/access_token/refresh", request, AuthQuestAccessTokenResponse.class)).thenReturn(response);

        RefreshTokenRequest refreshTokenRequest = new RefreshTokenRequest(fakeAccessToken);

        AuthQuestAccessTokenResponse actual = authService.getRefreshToken(refreshTokenRequest);

        verify(authRestTemplate).postForObject("/api/oauth/access_token/refresh", request, AuthQuestAccessTokenResponse.class);
        assertThat(actual).isEqualTo(response);

    }

    @Test
    public void shouldReturnTrueIfSpaceNameIsIncludedInScopes(){
        AuthQuestJWT jwt = new AuthQuestJWT(
                "",
                asList("spacetwo", "spacethree"),
                "",
                "",
                ""
        );

        boolean actual = authService.authenticateScope(jwt, "SpaceTwo");

        assertThat(actual).isTrue();
    }

    @Test
    public void shouldReturnFalseIfSpaceNameIsNotIncludedInScopes() {
        AuthQuestJWT jwt = new AuthQuestJWT(
                "",
                asList("spacetwo", "spacethree"),
                "",
                "",
                ""
        );

        boolean actual = authService.authenticateScope(jwt, "Space");

        assertThat(actual).isFalse();
    }


}

File: api/src/test/kotlin/com/ford/internalprojects/peoplemover/auth/AuthServiceTest.kt
package com.ford.internalprojects.peoplemover.auth

import com.ford.internalprojects.peoplemover.auth.exceptions.InvalidTokenException
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.mockito.Mockito.`when`
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.jwt.JwtDecoder
import org.springframework.security.oauth2.jwt.JwtException
import java.time.Instant

@SpringBootTest
class AuthServiceTest {
    @Autowired
    lateinit var authservice: AuthService

    @MockBean
    lateinit var jwtDecoder: JwtDecoder

    @Test
    fun `validateToken - should return token if valid in ADFS`() {
        val fakeJwt = createMockJwt(true)
        val authVerifyResponse = OAuthVerifyResponse(
            "USER_ID",
            emptyList(),
            fakeJwt.expiresAt!!.toEpochMilli(),
            "https://localhost",
            "USER_ID"
        )
        `when`(jwtDecoder.decode("token")).thenReturn(fakeJwt)

        val actual = authservice.validateToken("token")

        assertThat(actual).isEqualTo(authVerifyResponse)
    }

    @Test
    fun `validateToken - should throw token exception if not valid in any provider`() {
        val accessToken = "valid_Token"
        `when`(jwtDecoder.decode(accessToken)).thenThrow(JwtException("INVALID JWT"))

        assertThatThrownBy { this.authservice.validateToken(accessToken) }
            .isInstanceOf(InvalidTokenException::class.java)
    }
}

fun createMockJwt(isUserToken: Boolean): Jwt {
    val headers = HashMap<String, Any>()
    headers["typ"] = "JWT"
    val claims = HashMap<String, Any>()
    if (isUserToken) {
        claims["sub"] = "USER_ID"
    } else {
        claims["appid"] = "APP_ID"
    }
    val issuedAt = Instant.MIN
    val expiredAt = Instant.now()
    claims["expiresAt"] = expiredAt
    claims["iss"] = "https://localhost"
    return Jwt("token", issuedAt, expiredAt, headers, claims)
}
